rfcs/20211118-support-implicit-concat/README.md

```diff
@@ -0,0 +1,113 @@
+automatically - it must be enabled on user side manually due to several
+reasons. Let consider them on an example which includes only two tensors to
+create a bigger one:
+* It may happen that two inputs, being an output from other primitives, may have
+  different memory layout which are incompatible, e.g., aBx32b and ABx16a32b. In
+  this case reorder would be needed no matter what, thus, doesn't make sense to
+  avoid explicit concat which would execute such reorder as a part of operation.
+  It means that user oversees that formats are same or compatible for implicit
+  merging. Note that this holds for each architecture and even if input memory
+  formats are same for one architecture, it does not necessarily mean that the
+  other one would have exact format or match would happen.
+* But even though formats are same, it may happen that implicit concat is still
+  impossible, e.g., first input has padding. Despite in theory implicit concat
+  is still possible, in the reality it may bring only troubles since second
+  input would highly likely be unaligned in memory and may lead to various
+  "interesting" performance effects or, even, crashes, so that using explicit
+  concat would be at least safer, but at most more beneficial.
+
+Let's consider two scenarios that may happen when considering implicit concat:
+
+* Scenario 1, when concatenation should happen over the axis which is the
+  outermost dimension in memory. E.g., `nChw16c` inputs and output with `n`
+  equal to `1` and concatenating over channels `C`. In this simplified scenario,
+  two inputs are separate dense pieces, and primitives may be configured as
+  usual in this case. The only thing required is to provide a proper memory
+  handle for a second piece. In this case, everything can be enabled on user
+  side as of current oneDNN state, if using CPU raw pointers or GPU USM memory
+  kind. For OpenCL buffers this trick won't work, unfortunately.
+
+* Scenario 2, when concatenation should happen over the axis which is not the
+  outermost dimension in memory. This is a generic scenario. Same example but
+  for `n > 1` works here. Let's for simplicity take `n = 2`. In this case each
+  primitive which contributes to the final output would have non-trivial strides
+  between `n` points. oneDNN does not provide any API which would allow to user
+  initialize a memory descriptor with both blocking descriptor and strides
+  representing not dense case. But submemory API may be used to create such with
+  two stages: initialize original memory descriptor with tags coming from
+  compute primitive or passed from previous layer, and then perform submemory
+  of this descriptor from the concat destination. This would allow to achieve
+  the desired. This general scenario will utilize offset0 and strides in
+  implementations internally but it requires support from the library and it's
+  not fully under user's control now. But such usage is independent from memory
+  kind.
```

Comment:
It is mainly about creating data with implicit concat, i.e. submemory-as-an-output-buffer. As in the pooling example that I shared, it is necessary to support submemory from input as well.

Reply:
Thanks for noticing. I've added this case as a separate paragraph.